src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u9 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u10 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u11 }))
+           u0))
+           (err ERR-STACKING-UNREACHABLE))
+
+        ;; mark address in use over this range
+        (map-set pox-addr-reward-cycles
+            { pox-addr: pox-addr }
+            { first-reward-cycle: first-reward-cycle, num-cycles: num-cycles }
+        )
+
+        (ok true)
+    )
+)
+
+;; What is the minimum number of uSTX to be stacked in the given reward cycle?
+;; Used internally by the Stacks node, and visible publicly.
+(define-read-only (get-stacking-minimum (reward-cycle uint))
+    (let (
+        (ustx-stacked-so-far (get-total-ustx-stacked reward-cycle))
+    )
+    (begin
+        (if (< ustx-stacked-so-far (/ total-liquid-ustx u4))
+            ;; less than 25% of all liquid STX are stacked, so the threshold is smaller
+            (/ total-liquid-ustx STACKING-THRESHOLD-25)
+            ;; at least 25% of all liquid STX are stacked, so the threshold is larger
+            (/ total-liquid-ustx STACKING-THRESHOLD-100))
+    ))
+)
+
+;; Is the address mode valid for a PoX burn address?
+(define-private (check-pox-addr-version (version uint))
+    (and (>= version ADDRESS-VERSION-P2PKH)
+         (<= version ADDRESS-VERSION-P2WSH)))
+
+;; Is the given lock period valid?
+(define-private (check-pox-lock-period (lock-period uint)) 
+    (and (>= lock-period MIN-POX-REWARD-CYCLES) 
+         (<= lock-period MAX-POX-REWARD-CYCLES)))
+
+;; Register a PoX address for one or more reward cycles, and set how many uSTX it locks up initially.
+;; Will fail if the PoX address is already registered in one of the reward cycles.
+;; Will fail if the number of uSTX is beneath the lowest allowed Stacking threshold at the time of the call.
+(define-private (register-pox-addr-checked (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                                           (amount-ustx uint)
+                                           (first-reward-cycle uint)
+                                           (num-cycles uint)
+                                           (is-delegate bool))
+    (let (
+        (ustx-min (get-stacking-minimum first-reward-cycle))
+        (is-registered (is-pox-addr-registered pox-addr first-reward-cycle num-cycles))
+    )
+    (begin
+        ;; can't be registered yet if not a delegate
+        ;; (the delegate claims its PoX address when it registers)
+        (asserts! (or (not is-registered) is-delegate)
+            (err ERR-STACKING-POX-ADDRESS-IN-USE))
+
+        ;; minimum uSTX must be met
+        (asserts! (<= ustx-min amount-ustx)
```

Comment:
I don't think this can be evaluated at this point -- the threshold for participating in a reward cycle won't be known until that cycle's PoX anchor is chosen (this has to be the case, because the limit for inclusion is defined in terms of the participating stacks, and you cannot use a previous reward cycle's participation rate without risking inclusion of too many addresses in the event that participation goes up).

We _can_ set a lower bound of `(/ total-liquid-ustx 20000)`, but that still doesn't guarantee that a participant has "stacked enough" to get a reward address in a round.

Reply:
You're right -- the node will need to check at the start of the reward cycle which PoX addresses have locked enough STX.  But, that's not the point of this check.  The point of this check is a DDoS guard -- it limits the size of the reward cycle list that the node has to iterate through in order to get the reward set.  Otherwise, people could grief the system by locking up 1 uSTX and entering their burn address into the reward set map, thereby forcing the node to spend an attacker-controlled amount of time iterating through a long list of burn addresses that will never be considered due to their low lockup.

Reply:
Gotcha -- I think for the DoS-prevention, we'd want to use `(/ total-liquid-ustx 20000)`, rather than the dynamic check in `get-stacking-minimum`

Reply:
With this check, the maximum size of the reward set for a reward cycle can be 8,750 addresses -- the first 5,000 addresses each represent 1/20,000'th of the liquid STX (or, 25% of the liquid STX), and the next 3,750 addresses each represent 1/5,000th of the remaining liquid STX (or, 75% of the liquid STX).

The actual number of addresses to be paid out via PoX will be different, of course -- some addresses will be omitted entirely because they represent too-small of a lock-up, and some addresses will be need to be duplicated so they can be paid out more than once in the reward cycle.  But, my chief concerns are bounding the size of this Clarity data structure, and minimizing the amount of I/O that needs to happen to register a PoX address.  If there's a better way to achieve these two ends, please advise.

Reply:
I think the answer here is related to the discussion above about tiers -- as described in SIP-007, there aren't two tiers, but rather just a lower bound on limit scaling. Because of that, this check cannot strictly limit the set to 8750 addresses (rather than 20k addresses). If each caller of this function stacks with the "griefing lower bound" at that moment, the bound starts at 1/20,000 for the first 5000 addresses, but it does _not_ immediately increase to 1/5,000.

Reply:
Ah, yes, I see that now.  I'll alter this anti-DDoS check to use `(/ total-liquid-ustx 20000)`.

Reply:
See https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r483370419